### ISSUE TITLE
[QUESTION] Performance degradation issue

### ISSUE THREAD

**Issue by bisunny** [NVIDIA/Megatron-LM#543]
I used the latest version(branch 'main') and found some performance issues, There are some unclear synchronizations operations(repeat_interleave) here that lead to inefficiency in GPU utilization, as shown below, and these will not appear in older version.
![image](https://github.com/NVIDIA/Megatron-LM/assets/30709861/cf8df6d4-5a47-455f-8bf6-e51c4a4734d5)

**Comment by deepakn94** [comment]
What commit hash are you using? What command line?

**Comment by bisunny** [comment]
commit hash is 'f7727433293427bef04858f67b2889fe9b177d88'. The command line refers to the blog https://huggingface.co/blog/megatron-training,  I just used multiple gpus to distribute training and changed some parameters(like hidden_size,number_layer,seq_length) to make the model larger.

**Comment by bisunny** [comment]
> commit hash is 'f7727433293427bef04858f67b2889fe9b177d88'. The command line refers to the blog https://huggingface.co/blog/megatron-training, I just used multiple gpus to distribute training and changed some parameters(like hidden_size,number_layer,seq_length) to make the model larger.

**Comment by bisunny** [comment]
> Author

export CUDA_DEVICE_MAX_CONNECTIONS=1
export CUDA_VISIBLE_DEVICES=0,1,2,3
GPUS_PER_NODE=4
MASTER_ADDR=localhost
MASTER_PORT=6001
NNODES=1
NODE_RANK=0
WORLD_SIZE=$(($GPUS_PER_NODE*$NNODES))
DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE --nnodes $NNODES --node_rank $NODE_RANK --master_addr $MASTER_ADDR --master_port $MASTER_PORT"
CHECKPOINT_PATH=/mnt/workspace/flash_v2/codeparrot/checkpoint2
VOCAB_FILE=/mnt/workspace/flash_v2/codeparrot/vocab.json
MERGE_FILE=/mnt/workspace/flash_v2/codeparrot/merges.txt
DATA_PATH=/mnt/workspace/flash_v2/codeparrot_content_document
GPT_ARGS="--num-layers 40
--hidden-size 5120
--num-attention-heads 40
--seq-length 4096
--max-position-embeddings 4096
--micro-batch-size 2
--global-batch-size 8
--lr 0.0005
--train-iters 10000
--swiglu
--lr-decay-iters 150000
--lr-decay-style cosine
--lr-warmup-iters 2000
--weight-decay .1
--adam-beta2 .999
--bf16
--log-interval 10
--save-interval 3000
--eval-interval 3000
--eval-iters 3000
--use-flash-attn
--use-distributed-optimizer
--log-timers-to-tensorboard
--log-params-norm
--log-num-zeros-in-grad
--timing-log-level 2
--sequence-parallel
--overlap-grad-reduce
"

TENSORBOARD_ARGS="--tensorboard-dir experiments/tensorboard"
torchrun $DISTRIBUTED_ARGS \
        pretrain_gpt.py \
        --tensor-model-parallel-size 4 \
        --pipeline-model-parallel-size 1 \
        $GPT_ARGS \
        --vocab-file $VOCAB_FILE \
        --merge-file $MERGE_FILE \
        --save $CHECKPOINT_PATH \
        --data-path $DATA_PATH \
        $TENSORBOARD_ARGS > log 2>&1&

**Comment by bisunny** [comment]
commit hash is 'f7727433293427bef04858f67b2889fe9b177d88'. The command line refers to the blog https://huggingface.co/blog/megatron-training, I just used multiple gpus to distribute training and changed some parameters(like hidden_size,number_layer,seq_length) to make the model larger.

**Comment by deepakn94** [comment]
I think this should be resolved with dd74ea0b9a40b4dd5c8eacf8306bc0d63c94e54c.

**Comment by bisunny** [comment]
Thanks, I will try it

**Comment by bisunny** [comment]
@deepakn94  Thank you very much, the previous problem has been solved. And I have another question. When using gqa( self.num_attention_heads_per_partition // self.num_query_groups_per_partition > 1 = True, if training llama 2 70b model), the above situation will still occur (synchronous operation). Can the problem be fundamentally solved(If it is a problem)?

**Comment by deepakn94** [comment]
We are looking into this. Stay tuned.

**Comment by mayank31398** [comment]
@deepakn94 I have opened a PR for this ^^
Can you take a look?
Its working fine for me for a 3B parameter model on 8x H100s
I don't have access to A100s so can't test on those

I think this has been fixed with PyTorch 2.1, at least I am unable to replicate it

**Comment by mayank31398** [comment]
something similar is done in HF transformers here for LLama models: https://github.com/huggingface/transformers/blob/ae093eef016533a3670561fa9e26addb42d446d1/src/transformers/models/llama/modeling_llama.py#L237-L246

**Comment by mayank31398** [comment]
also, the repeat can be avoided when using flash 2, since Flash 2 does that itself
